File: src/main/java/io/scriptor/loader/IPackageEntry.java
package io.scriptor.loader;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public interface IPackageEntry {

    record ClassEntry(@NotNull Class<?> type) implements IPackageEntry {
    }

    record IterableEntry(@NotNull List<IPackageEntry> entries) implements IPackageEntry, Iterable<IPackageEntry> {

        public IterableEntry() {
            this(new ArrayList<>());
        }

        @Override
        public @NotNull Iterator<IPackageEntry> iterator() {
            return entries.iterator();
        }
    }

    record FutureEntry(@NotNull ResourceClassLoader classLoader, @NotNull String packageName)
            implements IPackageEntry {
    }
}

File: src/main/java/io/scriptor/loader/IPackageEntry.kt
package io.scriptor.loader

interface IPackageEntry {

    data class ClassEntry(val type: Class<*>) : IPackageEntry

    data class IterableEntry(val entries: MutableList<IPackageEntry>) : IPackageEntry, Iterable<IPackageEntry> {
        constructor() : this(ArrayList<IPackageEntry>())

        override fun iterator(): MutableIterator<IPackageEntry> {
            return entries.iterator()
        }
    }

    data class FutureEntry(val classLoader: ResourceClassLoader, val packageName: String) : IPackageEntry
}
